feat: simplify the fan-in/fan-out logic

AstraBert opened the pull request:
Closes #36 

With the introduction of the `dispatch` and `receive` methods, we simplify the logic behind  the current fan-in/fan-out (or map-reduce) approach, as detailed in the referenced issue

masci commented on src/workflows/context/context.py:
```diff
@@ -375,6 +376,27 @@ def send_event(self, message: Event, step: str | None = None) -> None:
 
         self._broker_log.append(message)
 
+    async def dispatch(self, events: list[Event]) -> None:
```
Do we need two methods with very different names to do pretty much the same thing? I would propose one of the following:
- Remove `send_event` and only rely on `dispatch`. Single events would be sent with `ctx.dispatch([ev])`. Alternatively, make `dispatch` take `Union[Event, Sequence[Event]]`. The overhead of keeping `dispatched_events` is negligible
- Rename this function to `send_events` so it's clear what it does compared to the existing `send_event`

logan-markewich replied:
The issue with renaming to send_event is this fn is async, the other is sync 🤔

adrianlyjak replied:
`send_events` seems like a reasonable solution, although the state mutation means that this is indeed a little special, so it's perhaps confusing to name them too closely. (the state mutation is also the cause of the async)

Another thought: I'm also perhaps somewhat concerned that this doesn't have a great continuity for how to fan out multiple times before joining. That's probably a rare scenario, just would be nice if we had a more composable pattern.

Like, I'm almost wondering if there's something we could do at the type level instead of checking for the receives? Probably not... it's maybe too dynamic (and this would be much more complicated).

But like, if you could somehow tag it as `events: All[SomeEventType, MaybeSomeOtherEventType]`, this could trigger the workflow step as pending as soon as one of the events is added to the queue, and then the step would only be run run once all of the selected types within the queue were resolved. And then `All` ends up being an annotated type for a list, so you'd have a list in the function.

Pie in the sky, but seemed worth mentioning.

masci commented on src/workflows/context/context.py:
```diff
@@ -375,6 +376,27 @@ def send_event(self, message: Event, step: str | None = None) -> None:
+        dispatched = cast(dict, await self.store.get("dispatched_events", default={}))
+
+        for event in events:
+            if type(event) in dispatched:
+                dispatched[type(event)] += 1
+            else:
+                dispatched.update({type(event): 1})
```
You can get rid of this using a `defaultdict`

logan-markewich commented on src/workflows/context/context.py:
```diff
@@ -375,6 +376,27 @@ def send_event(self, message: Event, step: str | None = None) -> None:
+                dispatched.update({type(event): 1})
+            self.send_event(event)
+
+        await self.store.set("dispatched_events", dispatched)
```
so, the default state is dictstate, you can set anything there magically :) 

But this actually raises another issue with this approach -- if the user sets a custom state model, this will actually fail unless their state model has dispatched_events in it

logan-markewich replied:
We might need a solution that doesn't abuse the state store.... (I'm realizing now that wait_for_event has the same bug 😬)

adrianlyjak replied:
you'll need to make sure it serializes with the context still though

adrianlyjak commented on docs/docs/understanding/workflows/concurrent_execution.md:
```diff
@@ -24,6 +24,29 @@ class ParallelFlow(Workflow):
+        await ctx.dispatch(events)
+
+    @step(num_workers=4)
+    async def step_two(self, ctx: Context, ev: StepTwoEvent) -> StopEvent:
```
won't this stop the workflow after the first event since this returns a `StopEvent`? Might be better to leave it open ended/incomplete rather than returning a stop event

adrianlyjak commented on docs/docs/understanding/workflows/concurrent_execution.md:
```diff
@@ -62,6 +85,42 @@ The `step_three` step is fired every time a `StepThreeEvent` is received, but `c
+        self, ctx: Context, ev: StepThreeEvent
+    ) -> StopEvent | None:
+        # wait until we receive 3 events
+        result = await ctx.receive(ev, type(ev))
```
nit: I'm not sure `receive` is the most intuitive name here for accumulating multiple events. What about `gather`?

Also, this might be too magical, but the bailing with `None` is kind of confusing. We could perhaps consider throwing a special exception from this call that the workflow infrastructure interprets the same as `None`. Like a `NotReady` exception or something

AstraBert commented:
So @masci I removed the EventsQueue class and renamed the methods: now they are called `send_events` and `receive_events`: I'm not super opinionated about the naming, I just think it makes sense to have the send-receive couple

masci commented:
@AstraBert I'm fine with `gather`, it's the two sets of methods doing basically the same thing that I think we should consolidate. I think having *only* `send_events/gather` is a good enough interface, a bit more annoying to use for single events but overall more obvious to use.

AstraBert commented:
Ok so let's get rid of send_event! :))

logan-markewich commented on src/workflows/context/context.py:
````diff
@@ -535,6 +536,66 @@ async def my_step(self, ctx: Context, ev: StartEvent) -> WorkerEvent | GatherEve
+        Examples:
+        ```
+        class MultipleEventsWorkflow(Workflow):
+        @step
````
nit: class method needs indent

masci commented on src/workflows/context/context.py:
```diff
@@ -480,17 +489,9 @@ async def synthesize(
-        Args:
-            message (Event): The event to enqueue.
-            step (str | None): Optional step name to target.
+    def _send_event(self, message: Event, step: str | None = None) -> None:
```
Let's keep this around as `send_event` and make it a shallow wrapper of `send_events`.

masci commented on src/workflows/context/context.py:
```diff
@@ -535,6 +536,66 @@ async def my_step(self, ctx: Context, ev: StartEvent) -> WorkerEvent | GatherEve
+            else:
+                self._send_event(event, step)
+
+    def receive_events(
```
I see in the decorator draft you're thinking about a `gather` parameter - let's call this function `gather_events` so the relation will be obvious.

masci commented on src/workflows/context/context.py:
```diff
@@ -414,7 +423,7 @@ def _get_event_buffer_id(self, events: list[Type[Event]]) -> str:
         # Fall back to creating a stable identifier from expected events
         return ":".join(sorted(self._get_full_path(e_type) for e_type in events))
 
-    def collect_events(
```
Let's keep this around and alias it to `gather_events` with a deprecation notice

masci reviewed (changes requested):
After the conversations we had offline I think it would be easier to keep `send_event` around and make it a wrapper of `send_events`, as well as making `collect_events` an alias for `gather_events`. This way we can make the upgrade path 1.x -> 2.0 a bit easier and avoid one deprecation warning in 1.3 (we still need a deprecation warning for `collect_events`).

logan-markewich commented on src/workflows/context/context.py:
````diff
@@ -535,6 +537,121 @@ async def my_step(self, ctx: Context, ev: StartEvent) -> WorkerEvent | GatherEve
+            ctx.collect_events(ev, [OutputAEvent, OutputBEvent])
+        ```
+        """
+        return self.gather_events(ev, expected)
````
if I'm understanding this correctly, `gather_events` has a different functionality compared to the old `collect_events()` API -- gather will check for all recorded events of a certain type and wait/collect them

the older `collect_events()` relies on the user to pass in how many to collect 

I think because if this we should call `self._collect_events()` here? What do you think?

AstraBert replied:
Done!

masci replied:
Correct me if I'm wrong, but I think `gather_events` can actually be a drop-in replacement for  `collect_events()`, while it's not the case the other way around.

This:
```py
events = ctx.collect_events(ev, [QueryEvent, RetrieveEvent])
# events is None if one of [QueryEvent, RetrieveEvent] wasn't sent
# events is [QueryEvent, RetrieveEvent] when both events are sent
```
is equivalent to this:
```py
events = ctx.gather_events(ev, [QueryEvent, RetrieveEvent])
# events is None if one of [QueryEvent, RetrieveEvent] wasn't sent
# events is [QueryEvent, RetrieveEvent] when both events are sent
```

The difference is `gather_events` can be used in a way that `collect_events` cannot. For example, if I have a step doing `send_events([QueryEvent, QueryEvent, QueryEvent])`, these are different:
```py
events = ctx.collect_events(ev, [QueryEvent])
# events returns [QueryEvent]

events = ctx.gather_events(ev, [QueryEvent])
# events returns [QueryEvent, QueryEvent, QueryEvent]
```

AstraBert replied:
Yeah, and now that I am thinking of it, if `send_event` is now a simple wrapper around `send_events`, it might make sense to use `gather_events` within the `collect_event` wrapper.

logan-markewich replied:
I'm specifically think about the case where you do

```
ctx.collect_events(ev, [SomeEvent]*10)
```

I think the above won't work if we are just wrapping gather_events.... going to add a unit test to confirm

logan-markewich replied:
ok we have tests for the above, and they work! (Although not sure how tbh lol)
